File: api_sample/csharp/rest-list-cameras/src/NxServerClient.cs
// Copyright 2018-present Network Optix, Inc. Licensed under MPL 2.0: www.mozilla.org/MPL/2.0/
// Log in to ONE Nx VMS server/site and list its cameras (C#).
//
// C# port of ../../python/rest-list-cameras and ../../node/rest-list-cameras, on
// the latest /rest/v4 API. Uses the built-in HttpClient + System.Text.Json — no
// third-party packages.
//
// This talks DIRECTLY to a single VMS server/site (not via the cloud). It is the
// step that actually lists cameras, which the Cloud CDB cannot do. The flow
// follows Network Optix's recommended bearer-token authentication:
//
//   1. Log in:   POST   /rest/v4/login/sessions  { username, password, setCookie:false }
//                -> { "token": ... }
//   2. List:     GET    /rest/v4/devices          (Authorization: Bearer <token>)
//   3. Log out:  DELETE /rest/v4/login/sessions/<token>   (release the session)
//
// "Devices" are the cameras (and other media devices) attached to the site.
//
// Connecting: the host is the server, e.g. https://192.168.1.10:7001 (note the
// https + port), or a cloud relay address like https://<siteId>.relay.vmsproxy.com.
// Local servers usually present a self-signed certificate, so for a lab server you
// will typically need --insecure.

using System.Net;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;

namespace NxListCameras;

public sealed class AuthException : Exception
{
    public AuthException(string message) : base(message) { }
}

public sealed class ApiException : Exception
{
    public ApiException(string message) : base(message) { }
}

/// <summary>One camera, trimmed to the fields the table shows.</summary>
public sealed record Camera(string Name, string Status, string Model, string Id);

public sealed class NxServerClient
{
    public const string Api = "/rest/v4";

    private readonly HttpClient _http;
    private readonly string _host;

    public string? Token { get; private set; }

    public NxServerClient(HttpClient http, string host)
    {
        _http = http;
        _host = host.TrimEnd('/');
    }

    public async Task<string> LoginAsync(
        string user, string password, CancellationToken cancellationToken = default)
    {
        string url = $"{_host}{Api}/login/sessions";
        var body = new Dictionary<string, object>
        {
            ["username"] = user,
            ["password"] = password,
            ["setCookie"] = false,
        };

        using var content = new StringContent(JsonSerializer.Serialize(body), Encoding.UTF8, "application/json");

        HttpResponseMessage response;
        try
        {
            response = await _http.PostAsync(url, content, cancellationToken);
        }
        catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException)
        {
            throw new ApiException($"Could not reach {url}: {ex.Message}");
        }

        using (response)
        {
            if (response.StatusCode is HttpStatusCode.Unauthorized or HttpStatusCode.Forbidden)
            {
                throw new AuthException(
                    $"Login unauthorized (HTTP {(int)response.StatusCode}). Check the "
                    + "username/password, and that you are using a local (not cloud) user.");
            }
            if (!response.IsSuccessStatusCode)
            {
                string text = await SafeReadAsync(response);
                throw new ApiException(
                    $"Login failed: HTTP {(int)response.StatusCode} {Truncate(text, 200)}");
            }
            string json = await response.Content.ReadAsStringAsync(cancellationToken);
            Token = ExtractToken(json);
            return Token;
        }
    }

    /// <summary>Use a bearer token obtained elsewhere (skip login).</summary>
    public void UseToken(string token) => Token = token;

    public async Task<IReadOnlyList<Camera>> ListCamerasAsync(CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrEmpty(Token))
        {
            throw new ApiException("Not logged in. Call LoginAsync() or UseToken() first.");
        }

        string url = $"{_host}{Api}/devices";
        using var request = new HttpRequestMessage(HttpMethod.Get, url);
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", Token);

        HttpResponseMessage response;
        try
        {
            response = await _http.SendAsync(request, cancellationToken);
        }
        catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException)
        {
            throw new ApiException($"Could not reach {url}: {ex.Message}");
        }

        using (response)
        {
            if (response.StatusCode is HttpStatusCode.Unauthorized or HttpStatusCode.Forbidden)
            {
                throw new AuthException(
                    $"Listing devices unauthorized (HTTP {(int)response.StatusCode}). Check the "
                    + "username/password, and that you are using a local (not cloud) user.");
            }
            if (!response.IsSuccessStatusCode)
            {
                string text = await SafeReadAsync(response);
                throw new ApiException(
                    $"Listing devices failed: HTTP {(int)response.StatusCode} {Truncate(text, 200)}");
            }

            string json = await response.Content.ReadAsStringAsync(cancellationToken);
            return NormalizeCameras(json);
        }
    }

    /// <summary>DELETE the session so the token cannot be reused. Best-effort.</summary>
    public async Task LogoutAsync(CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrEmpty(Token)) return;
        string url = $"{_host}{Api}/login/sessions/{Token}";
        try
        {
            using var request = new HttpRequestMessage(HttpMethod.Delete, url);
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", Token);
            using HttpResponseMessage _ = await _http.SendAsync(request, cancellationToken);
        }
        catch
        {
            // Logout is cleanup; never let it crash the program.
        }
        finally
        {
            Token = null;
        }
    }

    // -----------------------------------------------------------------------
    // Parsing helpers (pure = easy to test)
    // -----------------------------------------------------------------------

    public static string ExtractToken(string json)
    {
        JsonDocument doc;
        try { doc = JsonDocument.Parse(json); }
        catch (JsonException) { throw new ApiException("Login response was not valid JSON."); }
        using (doc)
        {
            if (doc.RootElement.ValueKind == JsonValueKind.Object
                && doc.RootElement.TryGetProperty("token", out JsonElement el)
                && el.ValueKind == JsonValueKind.String)
            {
                return el.GetString()!;
            }
            throw new ApiException("Login response did not contain a token.");
        }
    }

    /// <summary>
    /// The /rest/v4/devices body is sometimes a bare array and sometimes an
    /// envelope { "reply": [...] }. Unwrap to a list of Camera records.
    /// </summary>
    public static IReadOnlyList<Camera> NormalizeCameras(string json)
    {
        JsonDocument doc;
        try { doc = JsonDocument.Parse(json); }
        catch (JsonException) { throw new ApiException("Devices response was not valid JSON."); }

        using (doc)
        {
            JsonElement root = doc.RootElement;
            JsonElement array;
            if (root.ValueKind == JsonValueKind.Array)
            {
                array = root;
            }
            else if (root.ValueKind == JsonValueKind.Object
                     && root.TryGetProperty("reply", out JsonElement reply)
                     && reply.ValueKind == JsonValueKind.Array)
            {
                array = reply;
            }
            else
            {
                return Array.Empty<Camera>();
            }

            var cameras = new List<Camera>();
            foreach (JsonElement cam in array.EnumerateArray())
            {
                cameras.Add(new Camera(
                    Str(cam, "name"), Str(cam, "status"), Str(cam, "model"), Str(cam, "id")));
            }
            return cameras;
        }
    }

    private static string Str(JsonElement obj, string prop)
    {
        if (obj.ValueKind == JsonValueKind.Object && obj.TryGetProperty(prop, out JsonElement el))
        {
            return el.ValueKind == JsonValueKind.String ? el.GetString() ?? "" : el.ToString();
        }
        return "";
    }

    /// <summary>Render the cameras as a simple aligned text table.</summary>
    public static string FormatCamerasTable(IReadOnlyList<Camera> cameras)
    {
        if (cameras.Count == 0) return "No cameras found on this site.";

        var rows = new List<string[]> { new[] { "NAME", "STATUS", "MODEL", "ID" } };
        foreach (Camera c in cameras)
        {
            rows.Add(new[] { c.Name, c.Status, c.Model, c.Id });
        }
        int[] widths = new int[4];
        foreach (string[] row in rows)
        {
            for (int i = 0; i < 4; i++) widths[i] = Math.Max(widths[i], row[i].Length);
        }
        var sb = new StringBuilder();
        foreach (string[] row in rows)
        {
            var cells = new string[4];
            for (int i = 0; i < 4; i++) cells[i] = row[i].PadRight(widths[i]);
            sb.AppendLine(string.Join("  ", cells).TrimEnd());
        }
        return sb.ToString().TrimEnd('\n', '\r');
    }

    private static async Task<string> SafeReadAsync(HttpResponseMessage response)
    {
        try { return await response.Content.ReadAsStringAsync(); }
        catch { return string.Empty; }
    }

    private static string Truncate(string s, int max) => s.Length <= max ? s : s[..max];
}

File: api_sample/web/rest-list-cameras-local-browser/nx-server-client.mjs
// Copyright 2018-present Network Optix, Inc. Licensed under MPL 2.0: www.mozilla.org/MPL/2.0/
/**
 * Log in to ONE Nx VMS server and list its cameras — from the BROWSER.
 *
 * Browser/front-end counterpart of ../../node/rest-list-cameras and
 * ../../python/rest-list-cameras, on the latest /rest/v4 API. This is the
 * DIRECT-to-a-single-server variant: it talks to one configured VMS server
 * (not the cloud, no relay). This file holds the pure, framework-free logic so
 * it can be (a) imported by the page (app.mjs) and (b) tested offline with
 * node:test and a fake fetch — the same pattern every other sample uses.
 *
 * The token flow is identical to the Node/Python sample:
 *
 *   1. Log in:    POST {server}/rest/v4/login/sessions
 *                   { username, password, setCookie:false }  -> { token }
 *   2. List:      GET  {server}/rest/v4/devices   (Authorization: Bearer <token>)
 *   3. Log out:   DELETE {server}/rest/v4/login/sessions/<token>  (best-effort)
 *
 * WHY THE BROWSER ALWAYS GOES THROUGH THE PROXY (read the README):
 *
 *   A local Nx server is a different origin from this page and does NOT send
 *   CORS headers for it, so the browser blocks direct calls. It also usually
 *   presents a self-signed TLS certificate, which the browser refuses outright.
 *   The included proxy.mjs serves this page AND relays the calls same-origin
 *   (and can accept the self-signed cert with --insecure), so the client just
 *   uses one relative route:
 *
 *        {baseUrl}/server/...   -> the configured VMS server
 *
 *   baseUrl defaults to "" (same origin = the proxy that served the page), so
 *   there is nothing for the user to configure. The server's address is set on
 *   the proxy at start time (server.mjs --server-host ...), not on the page.
 */

// API version path segment. v4 is the latest Nx REST API.
export const API = "/rest/v4";

// ---------------------------------------------------------------------------
// Errors
// ---------------------------------------------------------------------------

export class AuthError extends Error {
  constructor(message) {
    super(message);
    this.name = "AuthError";
  }
}

export class ApiError extends Error {
  constructor(message) {
    super(message);
    this.name = "ApiError";
  }
}

// ---------------------------------------------------------------------------
// Client
// ---------------------------------------------------------------------------

export class NxServerClient {
  /**
   * @param {object} cfg
   * @param {string} cfg.user       Local server username.
   * @param {string} cfg.password
   * @param {string} [cfg.baseUrl]  Origin that serves the /server route.
   *        Defaults to "" = same origin (the proxy that served this page).
   *        You should not need to set this.
   * @param {function} [cfg.fetchImpl] Injected for offline tests. Defaults to
   *        the global fetch (browser or Node 18+).
   */
  constructor({ user, password, baseUrl = "", fetchImpl = null }) {
    this.user = user;
    this.password = password;
    this.baseUrl = (baseUrl || "").replace(/\/+$/, "");
    // Default to the global fetch, but call it through a wrapper so it keeps
    // its `window`/global receiver. Calling `this.fetchImpl(...)` where
    // fetchImpl IS window.fetch throws "Can only call Window.fetch on
    // instances of Window" in browsers — the bound wrapper avoids that.
    this.fetchImpl = fetchImpl || ((...args) => globalThis.fetch(...args));
    this.token = null;
  }

  /** Same-origin route the proxy forwards to the configured VMS server. */
  get serverUrl() {
    return `${this.baseUrl}/server`;
  }

  async login() {
    const url = `${this.serverUrl}${API}/login/sessions`;
    const body = {
      username: this.user,
      password: this.password,
      // A token (not a cookie) is what we want for a Bearer-header flow.
      setCookie: false,
    };

    let response;
    try {
      response = await this.fetchImpl(url, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(body),
      });
    } catch (exc) {
      throw new ApiError(
        `Could not reach the API at ${url}: ${exc.message}. ` +
          "Is the dev server running? (node server.mjs --server-host ...) — see the README.",
      );
    }
    if (response.status === 401 || response.status === 403) {
      throw new AuthError(
        `Login rejected (HTTP ${response.status}): the server did not accept this ` +
          "username/password. Verify they're a LOCAL server account (cloud users " +
          "use a different login flow).",
      );
    }
    if (!response.ok) {
      const text = await safeText(response);
      throw new ApiError(`Login failed: HTTP ${response.status} ${text.slice(0, 200)}`);
    }
    let data;
    try {
      data = await response.json();
    } catch {
      throw new ApiError("Login response was not valid JSON.");
    }
    this.token = data.token;
    if (!this.token) throw new ApiError("Login response did not contain a token.");
    return this.token;
  }

  _authHeader() {
    if (!this.token) throw new ApiError("Not logged in. Call login() first.");
    return { Authorization: `Bearer ${this.token}` };
  }

  async listCameras() {
    const url = `${this.serverUrl}${API}/devices`;
    let response;
    try {
      response = await this.fetchImpl(url, { headers: this._authHeader() });
    } catch (exc) {
      throw new ApiError(
        `Could not reach the server at ${url}: ${exc.message}. ` +
          "Is the dev server running? (node server.mjs --server-host ...) — see the README.",
      );
    }
    if (response.status === 401 || response.status === 403) {
      throw new AuthError(
        "The server rejected the token. Make sure the account has permission to " +
          "view the cameras on this site.",
      );
    }
    if (!response.ok) {
      const text = await safeText(response);
      throw new ApiError(`Listing devices failed: HTTP ${response.status} ${text.slice(0, 200)}`);
    }
    let data;
    try {
      data = await response.json();
    } catch {
      throw new ApiError("Devices response was not valid JSON.");
    }
    return normalizeCameras(data);
  }

  /** Delete the session token on the server. Best-effort cleanup. */
  async logout() {
    if (!this.token) return;
    const url = `${this.serverUrl}${API}/login/sessions/${this.token}`;
    try {
      await this.fetchImpl(url, { method: "DELETE", headers: this._authHeader() });
    } catch {
      // best effort
    } finally {
      this.token = null;
    }
  }
}

async function safeText(response) {
  try {
    return await response.text();
  } catch {
    return "";
  }
}

// ---------------------------------------------------------------------------
// Shaping the response for the UI
// ---------------------------------------------------------------------------

/**
 * The /rest/v4/devices response is sometimes a bare array and sometimes an
 * envelope { reply: [...] }. Unwrap to a plain array, then keep just the
 * fields the table renders.
 */
export function normalizeCameras(data) {
  let list = [];
  if (Array.isArray(data)) list = data;
  else if (data && typeof data === "object" && Array.isArray(data.reply)) list = data.reply;
  return list.map((cam) => ({
    name: cam.name ?? "",
    status: cam.status ?? "",
    model: cam.model ?? "",
    id: cam.id ?? "",
  }));
}

// ---------------------------------------------------------------------------
// Config: just the two things the user types. The server host is set on the
// proxy at start time, never on the page.
// ---------------------------------------------------------------------------

export function resolveConfig(values = {}) {
  const v = (key) => (values[key] === undefined || values[key] === null ? "" : String(values[key]).trim());
  return {
    user: v("user"),
    password: v("password"),
  };
}

export function missingFields(config) {
  return ["user", "password"].filter((k) => !config[k]);
}
